### PR TITLE
4220 attr normalizer attr export

### PR DESCRIPTION
**Pull request by a2xchip**
**Description (for Contributor and Core Developer)**

Extending list of attribute properties attribute normalizer normalize to full list. #4220

### CODE REVIEW

File: spec/Pim/Component/Connector/Normalizer/Flat/AttributeNormalizerSpec.php [hunk 1]
```diff
@@ -32,6 +32,25 @@ function let(
...
+
+        $attribute->getLocaleSpecificCodes()->willReturn(['en_US', 'fr_FR']);
+        $attribute->getMaxCharacters()->willReturn(null);
+        $attribute->getValidationRule()->willReturn(null);
+        $attribute->getValidationRegexp()->willReturn(null);
+        $attribute->isWysiwygEnabled()->willReturn(false);
+        $attribute->getNumberMin()->willReturn('');
+        $attribute->getNumberMax()->willReturn('');
+        $attribute->isDecimalsAllowed()->willReturn(false);
+        $attribute->isNegativeAllowed()->willReturn(false);
```
Comment (pierallard) [inline spec/Pim/Component/Connector/Normalizer/Flat/AttributeNormalizerSpec.php]: You duplicated isWysiwygEnabled, isDecimalsAllowed, isNegativeAllowed, getNumberMin and getNumberMax returns ;)

File: src/Pim/Component/Catalog/Normalizer/Structured/AttributeNormalizer.php [hunk 2]
```diff
@@ -153,4 +163,20 @@ protected function normalizeOptions(AttributeInterface $attribute)
 
         return $data;
     }
+
+    /**
+     * Normilize date option
+     * 
+     * @param \DateTime|null when null returns '' (empty string)
+     * 
+     * @return  string
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/AttributeNormalizer.php]: extra space between @return and string

File: src/Pim/Component/Catalog/Normalizer/Structured/AttributeNormalizer.php [hunk 3]
```diff
@@ -153,4 +163,20 @@ protected function normalizeOptions(AttributeInterface $attribute)
...
+    /**
+     * Normilize date option
+     * 
+     * @param \DateTime|null when null returns '' (empty string)
+     * 
+     * @return  string 
+     */
+    protected function normalizeDate($date = null)
+    {
+        if (!is_null($date) && $date instanceof  \DateTime) {
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/AttributeNormalizer.php]: Extra space too

**Comment by pierallard** [comment]
Hey @a2xchip Thanks for this PR :tada : Always a pleasure to get new features!
I just notice some checkstyle in your code. I will run the CI to give you results, but I think this test will raise because there is missing columns! features/export/export_attributes.feature
I hold you informed of the result.

**Comment by a2xchip** [comment]
@pierallard Hello! Nice to hear it. I didn't know that attribute export is checked via both phpspec and behat. And updated only specs.

File: spec/Pim/Component/Connector/Normalizer/Flat/AttributeNormalizerSpec.php (hunk 1 above)
Comment (a2xchip) [inline spec/Pim/Component/Connector/Normalizer/Flat/AttributeNormalizerSpec.php]: ))) Flat attribute normalizer extends Structured that was target.

**Comment by a2xchip** [comment]
@pierallard It is OK to fix them fast. How do you run behat? Is there selenium or other included via packagist?

**Comment by pierallard** [comment]
@a2xchip 
Yes, we have 2 ways for testing:
- the "specs" for testing functions
- the "behats" for testing the UI.
As export is accessible via the UI, there is behats tests for this.

So you want to correct behats, cool ! :-) Very sorry, our documenation about contributing and testing should be improved :sweat_smile:
Indeed, you need to
- download Selenium ( http://www.seleniumhq.org/download/ )
- run it (as deamon or in a standalone instance)
- configure a test database (copy paste https://github.com/akeneo/pim-community-dev/blob/master/app/config/parameters.yml.dist to parameters_test.yml and set your configuration)
- install the behat environement (`app/console pim:install --env=behat` or `--env=test`, I don't remember)
- configure a behat environment (copy paste https://github.com/akeneo/pim-community-dev/blob/master/behat.yml.dist to behat.yml, and set your own `base_url` (2 matching lines). These base_url will be accessible, it depends of your own install, maybe you have to create a new available site on apache and update your /etc/hosts)

Then, bin/behat thetest.feature !

Seeing all this, I realize how complex the process is... We definitely should improve it.

Feel free to ask if you encounter an issue.

**Comment by a2xchip** [comment]
@pierallard In my projects using behat i use to include these: 

* behat/behat
* behat/mink
* behat/mink-extension
* behat/mink-goutte-driver
* behat/mink-selenium2-driver
* peridot-php/webdriver-manager

```peridot-php/webdriver-manager``` downloads selenium first time and it may update it to latest version
```bin/manager start``` starts selenium

We may implement own package or simply include this and create wrapper over it. What do other developers think about it? @nidup

**Comment by a2xchip** [comment]
Updates to PR branch. There shouldn't be more problems i think )))

**Comment by nidup** [comment]
Thank you @a2xchip! (I run the PR on our Jenkins CI, still a small glitch that I'll fix just after the merge).

In fact we don't run behat on travis ci because we have too much scenario (~900 in CE) and it takes too much time. We have a custom Distributed CI based on Jenkins which split a build and each scenario is launched by one of our 200 VPS. Last time I tried to run all behat on my laptop, it took ~15h, our CI gives us an answer in ~25 min (we continuously improve this system). We would especially improve it to allow to contributors to be able to launch builds or even better to automatically run builds and give results directly on Github.

I didn't know peridot-php/webdriver-manager, it looks interesting, imho we could discuss about it in the context of the "how to improve contributing doc".

:tada: I merge the PR, welcome in the contributors team :tada:

**Comment by nidup** [comment]
Hi @a2xchip !

Hope everything is ok for you!

FYI, we've added new badges in badger, you can claim them if you want http://badger.akeneo.com/badge/41acec2c-649f-11e6-92dc-d60437e930cf and http://badger.akeneo.com/badge/0a25c62e-1bae-11e6-92dc-d60437e930cf

NB: we're adding more of them in the upcoming weeks 

Have a good day!

**Pull request merged**